neps/nep-0509.md

```diff
@@ -12,58 +12,107 @@ LastUpdated: 2023-09-19
+The early design of phase 2 relied on the security assumption that as long as there is one honest validator or fisherman tracking a shard, the shard is secure; by doing so, it naturally relied on protocol's ability to handle challenges (when an honest validator or fisherman detects a malicious behavior and submits a proof of such), state rollbacks (when validators agree that the submitted challenge is valid), and slashing (to punish the malicious validator). While it sounds straightforward and simple on paper, the complex interactions between these abilities and the rest of the protocol led to concrete designs that were extremely complicated, involving several specific problems we still don't know how to solve.
 
-As a result, the team sought alternative approach and concluded that stateless validation is the most realistic and promising one; stateless validation approach does not assume the existence of fishermen and assumes that a shard is secure if every single chunk in that shard is validated by a randomly sampled subset of all validators.
+As a result, the team sought alternative approaches and concluded that stateless validation is the most realistic and promising one; the stateless validation approach does not assume the existence of a fishermen, does not rely on challenges, and never rolls back state. Instead, it relies on the assumption that a shard is secure if every single chunk in that shard is validated by a randomly sampled subset of all validators, to always produce valid chunks in the first place.
 
 ## Specification
 
 ### Assumptions
 * In memory trie is enabled - [REF](https://docs.google.com/document/d/1_X2z6CZbIsL68PiFvyrasjRdvKA_uucyIaDURziiH2U/edit?usp=sharing)
-* State sync is enabled
-* Merkle Patricia Trie is underlying data structure
+* State sync is enabled (so that nodes can track different shards across epochs)
+* Merkle Patricia Trie continues to be the state trie implementation
 * TBD
 
 ### High level requirements
-* Block processing time should not take more than what it takes today.
-* Additional load on network and node should not affect other functionalities.
-* Security of protocol must not degrade.
-  * Validator assignment for both chunk validation and block validation should not create any security vulnerability.
-* Validators should be rewarded with the same amount as they are now.
 * No validator needs to track all shards.
-* Only chunk producers need to maintain state locally.
-* State witness size should be small enough to delivered over network.
-* Majority of undercharging issues should be mitigated.
-* Validator roles are updated accordingly to reflect the necessary changes with stateless validation.
-* [TBD] Resharding should work as expected after stateless validation in place.
+* Security of protocol must not degrade.
+  * Validator assignment for both chunk validation and block validation should not create any security vulnerabilities.
```

Comment:
I think the old point "Moreover, majority of undercharging issues should be mitigated." deserves a point here. Because it is not tracking all shards what allows in-memory trie to be used

Reply:
Any reason we are removing the following statement?

> Only chunk producers need to maintain state locally.

Reply:
These are mostly technicalities :)

I removed the undercharging point because it's not a "high level requirement". It's more of a good-to-have side effect. It's not a requirement for stateless validation to solve "a majority of undercharging issues" (which isn't even well defined).

For "only chunk producers need to maintain state locally", I also don't think this is a requirement. Even if everybody is required to track some state (and it's not all the state because there is already a point saying no validators needs to track all shards), stateless validation is still successful. If we're talking about the desire that chunk validators who don't also act as chunk producers do not need to track any shard at all, I think the motivation there is that they don't need to have beefy machines to track any state, but that is also covered by the point "Any additional load on network and compute should not negatively affect existing functionalities of *any* node in the blockchain", and that the cost of these should be acceptable. Also, "only chunk producers need to maintain state locally" seems to be a particular solution rather than a requirement, as we're also changing the definition of chunk producers.